File: public/ps/PS2/GraphADT/GraphADT/Graph.m
#import "Graph.h"

@implementation Graph

@synthesize isDirected;

// Checks that the representation invariant holds.
-(void)checkRep{
    // 5 points

}

- (id)initWithIsDirected:(BOOL)d {
    // MODIFIES: self
    // EFFECTS: returns a new Graph object g with g.isDirected = d
    // 5 points
    
    return nil; //Remove this line in actual implementation
}

- (void)addNode:(Node*)n {
    // REQUIRES: self != nil and n != nil
    // MODIFIES: self
    // EFFECTS: adds the node n to self
    // 5 points

    return; //Remove this line in actual implementation
}

- (void)removeNode:(Node*)n {
    // REQUIRES: self != nil and n != nil
    // MODIFIES: self
    // EFFECTS: removes the node n and all respective edges from self
    // 5 points

    return; //Remove this line in actual implementation
}

- (BOOL)containsNode:(Node*)n {
    // REQUIRES: self != nil and n != nil
    // EFFECTS: returns true if the node n is present in self
    // 5 points

    return nil; //Remove this line in actual implementation
}

- (void)addEdge:(Edge*)e {
    // REQUIRES: self != nil and e != nil
    // MODIFIES: self
    // EFFECTS: adds the edge e to self
    // 5 points

    return; //Remove this line in actual implementation
}

- (void)removeEdge:(Edge*)e {
    // REQUIRES: self != nil and e != nil
    // MODIFIES: self
    // EFFECTS: removes the edge e from self
    // 5 points

    return; //Remove this line in actual implementation
}

- (BOOL)containsEdge:(Edge*)e {
    // REQUIRES: self != nil and e != nil
    // EFFECTS: returns true if the edge e is present in self
    // 5 points

    return nil; //Remove this line in actual implementation
}

- (NSArray*)getEdgesFromNode:(Node*)sn toNode:(Node*)en {
    // REQUIRES: self != nil, sn != nil and en != nil
    // EFFECTS: returns the edges between sn and en if present,
    //          else, returns nil
    // 5 points

    return nil; //Remove this line in actual implementation
}

- (NSArray*)getAdjacentNodesForNode:(Node*)n {
    // REQUIRES: self != nil and n != nil
    // EFFECTS: returns all the nodes adjacent to n in self
    // 5 points

    return nil; //Remove this line in actual implementation
}

- (NSArray*)getAllNodes {
    // REQUIRES: self != nil
    // EFFECTS: returns all the nodes in self
    // 5 points

    return nil; //Remove this line in actual implementation
}

- (NSArray*)getAllEdges {
    // REQUIRES: self != nil
    // EFFECTS: returns all edges in self
    // 5 points

    return nil; //Remove this line in actual implementation
}

- (NSArray*)getShortestPathFromNode:(Node*)sn toNode:(Node*)en {
    // REQUIRES: self != nil, sn != nil and en != nil
    // EFFECTS: returns the shortest path from sn to en in self
    // 20 points

    return nil; //Remove this line in actual implementation
}

@end

File: public/ps/PS2/GraphADT/GraphADT/Graph.swift
/*
    `Graph` represents a graph (unsurprisingly!) More specifically, `Graph` 
    should be able to represent the following graph types with corresponding
    constraints:
        - Undirected graph
            + An undirected edge is represented by 2 directed edges
        - Directed graph
        - Simple graph
        - Multigraph
            + Edges from the same source to the same destination should have
            different cost
        - Unweighted graph
            + Edges' weights are to set to 1.0
        - Weighted graph

    Hence, the representation invariants for every Graph g:
        - g is either directed or undirected
        - All nodes in g must have unique labels
        - Multiple edges from the same source to the same destination must
        not have the same weight

    Finally, just like `Node` and `Edge`, `Graph` is a generic type with a type
    parameter `T` that defines the type of the nodes' labels.
*/


class Graph<T: Equatable> {
    typealias N = Node<T>
    typealias E = Edge<T>
    
    let isDirected: Bool
    
    //  5 points
    //  Construct a direct or undirected graph.
    init(isDirected: Bool) {
        
        self.isDirected = false // Please remove this line in actual implementation.
    }
    
    //  5 points
    //  Add `addedNode` to the graph. If `addedNode` already exists
    //  in the graph, do nothing.
    func addNode(addedNode: N) {

    }
    
    //  5 points
    //  Remove `removedNode` from the graph. If `removedNode` does
    //  not exist in the graph, do nothing.
    func removeNode(removedNode: N) {

    }
    
    //  5 points
    //  Return true if `targetNode` exists in the graph.
    func containsNode(targetNode: N) -> Bool {
        
        return false // Please remove this line in actual implementation.
    }
    
    //  5 points
    //  Add `addedEdge` to the graph. If `addedEdge` already exists,
    //  do nothing. If any of the nodes referenced in `eddedEdge` does
    //  not exist, add it to the graph.
    func addEdge(addedEdge: E) {
    
    }
    
    //  5 points
    //  Remove `removedEdge` from the graph. If `removedEdge` does not
    //  exist, do nothing.
    func removeEdge(removedEdge: E) {

    }
    
    //  5 points
    //  Return true if `targetEdge` exists in the graph.
    func containsEdge(targetEdge: E) -> Bool {
        
        return false // Please remove this line in actual implementation.
    }
    
    //  5 points
    //  Return edges directed from `fromNode` to `toNode`.
    func edgesFromNode(fromNode: N, toNode: N) -> [E] {
        
        return [E]() // Please remove this line in actual implementation.
    }
    
    //  5 points
    //  Return adjacent nodes of the `fromNode` i.e. there
    //  is an directed edge from `fromNode` to its adjacent node.
    func adjacentNodesFromNode(fromNode: N) -> [N] {
        
        return [N]() // Please remove this line in actual implementation.
    }
    
    //  5 points
    //  A read-only computed property that contains all the nodes 
    //  of the graphs.
    var nodes: [N] {
        
        return [N]() // Please remove this line in actual implementation.
    }
    
    //  5 points
    //  A read-only computed property that contains all the edges 
    //  of the graphs.
    var edges: [E] {
        
        return [E]() // Please remove this line in actual implementation.
    }
    
    //  5 points
    //  Check the representation invariants.
    private func _checkRep() {
        
    }
    
    //  20 points
    //  Return the shortest path from `fromNode` to `toNode`
    //  assuming that the graph is not a multigraph.
    func shortestPathFromNode(fromNode: N, toNode: N) -> [E] {
        
        return [E]() // Please remove this line in actual implementation.
    }
}
